### PR TITLE
tests: provide tests for emcute

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This provides a test application for `emcute` based on the MQTT-SN layer for `scapy` I provided in #11822. Since it uses UDP sockets it can be run without root, but requires `scapy` nonetheless.

~~Attached to this PR are fixes that I found with these tests.~~

#### ~~Fixes~~
- ~~**emcute: fix payload copy error for emcude_pub**
  `len` is used with the `memcpy()` to copy the payload to `tbuf`. With a
    payload provided that is just long enough to fill `tbuf`, `len += 6`
    leads to the `memcpy()` overriding data after `tbuf` (e.g. the
    `mutex` that is unlocked right after) and thus resulting in potential
    segmentation faults.
    Additionally `+ 6` can only be applied if the total packet length is
    below 256 (see spec), so `len + pos` is what needs to be provided to the
    corresponding send functions instead (`pos` adapts to the header length
    of the PUBLISH message)~~ (now in #11957)
- ~~**emcute: fix length field calculation**
    The length field in an MQTT packet carries the _total_ length of the
    packet. If it is below 256 (i.e. fits in one byte) only one byte is
    used for the length field. If it is larger than that 3 bytes are used,
    with the first byte having the value `0x01` and the remaining bytes
    representing the length in as a 2 byte unsigned integer in network byte
    order. Resulting from that it can be assessed that the check in
    `emcutes`'s `set_len()` function is wrong as it needs to be checked if
    `len` is lesser or equal to `0xff - 1`. `len <= (0xff - 1)` can be
    simplified to `len < 0xff`. For some larger packages this safes 2 bytes
    of wasted packet space.~~ (now in #11958)

### Testing procedure
Make sure you have a recent version of `scapy` for `python3` installed.

Run

```
BOARD="<your choice>" make -C all test
```

(the blacklisted boards are excluded of course).

### Issues/PRs references
Requires on #11822.

### CODE REVIEW

**Comment by miri64** [comment]
Ok, some details on the test server now that I cleaned it up a bit: it is a [`scapy` automaton](https://scapy.readthedocs.io/en/latest/advanced_usage.html#automata) (basically a deterministic automaton with sockets and packet parsing ability) and is defined as follows. 

mqttsn_test_server

You can generate this graphic yourself, using the following within a reachable portion of the test script:
```py
    MQTTSNServer.graph(options="-Gratio=.4", format="png",
            target="mqttsn_test_server.png")
```

This allows me to test some expected packet exchanges and include also some unexpected data exchanges. If the automaton reaches an error state, an exception is thrown. If it reaches the END state, the automaton's `run()` method stops.

**Comment by miri64** [comment]
I split out the two bug fixes into their own PRs.

**Comment by miri64** [comment]
I rebased and squashed to current master, since I closed #11822 due to `scapy` 2.4.3 being out (which includes that submodule). I also adopted the tests to include the submodule properly. Lastly, I removed the two bug fixes, as I moved them out into their own PRs, as I mentioned earlier.

**Comment by miri64** [comment]
Adapted for #12382.

**Comment by miri64** [comment]
Rebased to current master to incorporate latest fixes.

File: tests/emcute/Makefile [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                             i-nucleo-lrwan1 lsn50 maple-mini mega-xplained \
+                             microbit msb-430 msb-430h nrf51dongle nrf6310 \
+                             nucleo-f042k6 nucleo-f031k6 nucleo-f030r8
+                             nucleo-f070rb nucleo-f072rb nucleo-f103rb
+                             nucleo-f303k8 nucleo-f334r8 nucleo-l053r8
+                             nucleo-l031k6 nucleo-l073rz opencm904 \
+                             pic32-clicker pic32-wifire saml10-xpro \
+                             saml11-xpro spark-core stm32l0538-disco \
+                             stm32f0discovery thingy52 telosb waspmote-pro
+                             wsn430-v1_3b wsn430-v1_4 yunjia-nrf51822 z1
```
Comment (benpicco) [inline tests/emcute/Makefile]: goes to `Makefile.ci` now.

**Comment by miri64** [comment]
Rebased and squashed to current master, left `Makefile.ci` out for now, to get a more current list of files.

File: tests/emcute/main.c [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    if (_gw.port == 0) {
+        _gw.port = MQTTSN_DEFAULT_PORT;
+    }
+    if (argc >= 4) {
+        topic = argv[2];
+        message = argv[3];
+        len = strlen(message);
+    }
+
+
```
Comment (benpicco) [inline tests/emcute/main.c]: Vera thinks two consecutive empty lines are too much - I think that should be demoted to a warning (see #13400)

Comment (smlng) [reply tests/emcute/main.c]: why, what's the benefit of 2 empty lines here? I'd rather remove one here.

**Comment by miri64** [comment]
> Rebased and squashed to current master, left `Makefile.ci` out for now, to get a more current list of files.

Updated and also fixed some static errors and required adaptations for current master.

**Comment by miri64** [comment]
Test failure [unrelated](https://ci.riot-os.org/RIOT-OS/RIOT/11823/5507b1b4a2840d81c69af4c692a33fb23f1f84d5/output/run_test/examples/micropython/samr21-xpro:gnu.txt).

**Comment by benpicco** [comment]
I wonder what's wrong with the micropython test. It keeps failing for `samr21-xpro` quite consistently.
When I run it locally, it succeeds.

**Comment by miri64** [comment]
Maybe its not the test, but the test + a certain worker within Murdock that fails. The set of test runner workers is (due to the hardware requirements) quite small.

**Comment by miri64** [comment]
Murdock is happy now.

**Pull request merged**